Move CacheAccessor out of msal and fix up

element-of-surprise opened the pull request:
- Move CacheAccessor from msal/ to msal/cache
- msal.CacheAccessor -> cache.Token
- CacheAccssor.Before/After to IntoCache/ExportCache()
- storage.Manager.Serialize/Deserialize() to Marshal/Unmarshal()

This does a few things to help us out.  It moves this external interface into its own package that defines its scope. We move to more common naming, Marshal/Unmarshal().  The CacheAccessor(now cache.Token) has simplified methods that are clear using interfaces that give a single method call.  We also don't waste doing string/[]byte conversions that were not needed.  As a side effect, also shortens some test names.

keegan-caruso commented on msal/client_application.go:
```diff
@@ -11,16 +11,17 @@ import (
+func (n noopCacheAccessor) ExportCache(cache cache.Marshaler) {}
 
-// cache provides an internal cache. It is defined to allow faking the cache in tests.
+// managet provides an internal cache. It is defined to allow faking the cache in tests.
```
nit: typo "manager"

keegan-caruso commented on internal/storage/storage.go:
```diff
@@ -26,7 +26,7 @@ import (
+// updated on read/write calls. Unmarshal() replaces all data stored here with whatever
 // was given to it on each call.
 type Manager struct {
 	// TODO(jdoak): Going to refactor this (next PR).
```
is this up to date?

element-of-surprise replied:
It is, comment removed.

keegan-caruso commented on internal/storage/storage.go:
```diff
@@ -384,23 +384,23 @@ func (m *Manager) update(cache *Contract) {
 
 // TODO(jdoak): Change this to return []byte, not string.
```
this looks resolved

jhendrixMSFT commented on internal/storage/storage.go:
```diff
@@ -384,23 +384,23 @@ func (m *Manager) update(cache *Contract) {
-	serializedCache, err := json.Marshal(cache)
+// Marshal implements cache.Marshaler.
+func (m *Manager) Marshal() ([]byte, error) {
+	b, err := json.Marshal(m.Contract())
```
Can we not simply `return json.Marshal(m.Contract())` here?

element-of-surprise replied:
We can since we did the conversion from string returns to bytes, good catch.

keegan-caruso commented on msal/client_application.go:
```diff
@@ -56,12 +57,12 @@ func (client *clientApplication) acquireTokenSilent(ctx context.Context, silent
-	if sm, ok := client.cache.(*storage.Manager); ok {
-		client.cacheAccessor.BeforeCacheAccess(sm)
-		defer client.cacheAccessor.AfterCacheAccess(sm)
+	if sm, ok := client.manager.(*storage.Manager); ok {
```
Is there a reason why we don't want to call these on the fake?

element-of-surprise replied:
Skipping the first poinrt and straight to your second point, now that we have Unmarshaller/Marshaller, we could make this more grandular.  Introduced a Serializer interface, it now tests for that and if it is implemented it uses it.

keegan-caruso commented on msal/cache/cache.go:
```diff
@@ -0,0 +1,32 @@
+// opaque.
+type Token interface {
+	// IntoCache reads the cache in external storage into the internal cache, replacing it.
+	IntoCache(cache Unmarshaler)
```
I find this name to be confusing

jhendrixMSFT replied:
Agreed.  Also see usage in `confidential_client_application.go`

element-of-surprise replied:
Roger: would you have suggestions that you think are appropriate?  My mind is a little blank at the moment (I'll think about alternatives in the morning).

sangonzal replied:
Some ideas: 
- ImportCache, ExportCache
- ReadCache, WriteCache

keegan-caruso commented on internal/storage/storage.go:
```diff
@@ -4,7 +4,7 @@
 // Package storage holds all cached token information for MSAL. This storage can be
 // augmented with third-party extensions to provide persistent storage. In that case,
 // reads and writes in upper packages will call Serialize() to take the entire in-memory
```
"Serialize()" => "Marshall()"

keegan-caruso commented on test/devapps/sample_cache_accessor.go:
```diff
@@ -26,18 +26,18 @@ func (accessor *SampleCacheAccessor) BeforeCacheAccess(cache *storage.Manager) {
 	if err != nil {
 		log.Fatal(err)
 	}
-	err = ioutil.WriteFile(accessor.file, []byte(data), 0644)
```
we should fix this or suppress it

element-of-surprise replied:
I'm assuming you mean the log.Fatal().  I changed these to log.Error.  In reality, there is not much to be done when these happen other than ignore them or log them.  Really up to the developer on what they want to do.

keegan-caruso replied:
Sorry, missing context. gosec linter is failing. If you want to suppress it golangci-lint suppresses with a syntax of

// nolint:gosec

directly above the offending line.

otherwise permissions need to be at 600 or less

element-of-surprise commented:
Its old code, but it flagged it because I removed the string() conversion.  I changed it to 600, no reason not too I can see.

John Doak
Manager, Change Management SRE
jdoak@microsoft.com
________________________________
From: Keegan Caruso <notifications@github.com>
Sent: Tuesday, December 8, 2020 6:09 PM
To: AzureAD/microsoft-authentication-library-for-go <microsoft-authentication-library-for-go@noreply.github.com>
Cc: John Doak <John.Doak@microsoft.com>; Author <author@noreply.github.com>
Subject: Re: [AzureAD/microsoft-authentication-library-for-go] Move CacheAccessor out of msal and fix up (#102)


@keegan-caruso commented on this pull request.

________________________________

In test/devapps/sample_cache_accessor.go<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzureAD%2Fmicrosoft-authentication-library-for-go%2Fpull%2F102%23discussion_r538950065&data=04%7C01%7CJohn.Doak%40microsoft.com%7Ca63b5ceeb84f4158cf4b08d89be78154%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637430765887584005%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=Y14501ZIRXZNGNHQ35%2BjBBheqKNdyfVDarSvzCV9UlI%3D&reserved=0>:

>       if err != nil {
                log.Fatal(err)
        }
-       err = ioutil.WriteFile(accessor.file, []byte(data), 0644)


Sorry, missing context. gosec linter is failing. If you want to suppress it golangci-lint suppresses with a syntax of

// nolint:gosec

directly above the offending line.

otherwise permissions need to be at 600 or less

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzureAD%2Fmicrosoft-authentication-library-for-go%2Fpull%2F102%23discussion_r538950065&data=04%7C01%7CJohn.Doak%40microsoft.com%7Ca63b5ceeb84f4158cf4b08d89be78154%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637430765887584005%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=Y14501ZIRXZNGNHQ35%2BjBBheqKNdyfVDarSvzCV9UlI%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKCSL7DPVEQYTTHVQSUXJMLST3L6XANCNFSM4USSLIDQ&data=04%7C01%7CJohn.Doak%40microsoft.com%7Ca63b5ceeb84f4158cf4b08d89be78154%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637430765887593966%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=nJm9SbCMkp9pykx8YJW4UjSUgnx%2Fpwo4hcyuet1Itu0%3D&reserved=0>.

sangonzal commented on msal/cache/cache.go:
```diff
@@ -0,0 +1,38 @@
+
+// Token is used to provide external storage of token data. The data being passed is considered
+// opaque.
+type Token interface {
```
Token also doesn't seem quite right here. Token, in the context of MSAL has a clear meaning - A JWT with claims used for authentication/authorization. It also doesn't have a "er" suffix, which I I think is Go idiom and which is used elsewhere. 

This type is something completely unrelated to the above definition - is allows a user to set callbacks for marshalling/unmarshalling of the in-memory token cache. I can't think of a better name myself, but if I was to choose the previous name "accessor" seems more appropriate

element-of-surprise replied:
As to the Go idiom: "er" is only added to interface types that have a single method defined.  So instead of having Read.Read(), you have Reader.Read().  This does not apply on multi-method interfaces.

Do you feel the same way knowing that it is always:  cache.Token?   
We could just go all the way:  cache.ReadReplace
